### ISSUE TITLE
Docker should not update FORWARD chain on startup

### ISSUE THREAD

**Issue by jcstover** [moby/moby#23987]
We want to use docker in a restricted setup where we can controle who has access to the running containers. On startup docker adds rules to the FORWARD chain, empties the DOCKER chain and resets the DOCKER-ISOLATION chain.

Since we want to restrict access to the used bridges we want to add rules preventing this in the FORWARD chain, but the problem with this is that docker puts its own rules before our custom rules, in effect offering access to the containers for everyone.

Is there any setting to prevent docker from changing the FORWARD chain on startup and only update the DOCKER chain? or is it possible to implement such a switch. This way you can control access to the docker containers without messing up docker automatic firewall settings when using portmappings.

The --iptables=false setting works, but when starting containers nothing is added to the iptables anymore. So that isn't a workable solution either.. 

**Output of `docker version`:**

```
Client:
 Version:      1.11.1
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   5604cbe
 Built:        Wed Apr 27 00:34:42 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.11.1
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   5604cbe
 Built:        Wed Apr 27 00:34:42 2016
 OS/Arch:      linux/amd64

```

**Output of `docker info`:**

```
Containers: 2
 Running: 1
 Paused: 0
 Stopped: 1
Images: 111
Server Version: 1.11.1
Storage Driver: devicemapper
 Pool Name: docker-253:1-150997978-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: xfs
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 11.77 GB
 Data Space Total: 107.4 GB
 Data Space Available: 7.219 GB
 Metadata Space Used: 15.33 MB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.132 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 WARNING: Usage of loopback devices is strongly discouraged for production use. Either use `--storage-opt dm.thinpooldev` or use `--storage-opt dm.no_warn_on_loop_devices=true` to suppress this warning.
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.107-RHEL7 (2016-06-09)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: overlay bridge null host
Kernel Version: 3.10.0-327.13.1.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 7.79 GiB
Name: xxxx.xxxxxxxxxxx.xx
ID: H2Q6:YFIW:O4SQ:55KY:ZDPS:S5OT:XXJJ:HYEV:UDON:5TWF:3CMK:GILX
Docker Root Dir: /var/lib/docker
Debug mode (client): false
Debug mode (server): false
Http Proxy: http://192.168.202.13:3128
Https Proxy: http://192.168.202.13:3128
No Proxy: /var/run/docker.soc,localhost,127.0.0.1
Registry: https://index.docker.io/v1/
Labels:
 nl.bzk.rol=db
Cluster store: etcd://192.168.197.230:2379/cluster01
Cluster advertise: 192.168.197.231:2375
```

**Comment by cpuguy83** [comment]
Closing as this is a dup of #14041
If you could post your case over there, that'd be awesome!

Thanks!

**Comment by jcstover** [comment]
If the reported bug #24848 is solved this issue can also be closed. Both are trying to find a way to restrict access to the docker containers. Since the DOCKER chain is allways reset after the daemon is restarted I tried to modify the FORWARD chain, but also that will not work as mention in this issue.

**Comment by jcstover** [comment]
Finally found a solution found on this forum. On centos I changed the default iptabels file (/etc/sysconfig/iptables) to create a custom chain for restricting access to the docker containers and insert (not append) a rule to the FORWARD chain. Since it is an Insert it stays on top (DOCKER-ISOLATAIN will be first) but it allows filtering traffic to the docker containers

`:DOCKER-BLOCK - [0:0]
-I FORWARD -o docker_gwbridge -j DOCKER-BLOCK
-I FORWARD -o docker0 -j DOCKER-BLOCK
` 
After restart the FORWARD rules stays on top

**Comment by thaJeztah** [comment]
Do you think it's worth adding that to the documentation?

**Comment by jcstover** [comment]
Seen multiple issues on restricting access to the docker containers it can be usefull to add to the documentation. It should be noted that it is a sort of work-around because in the documentation says that a negated rule must be inserted in the DOCKER chain. But that chain is cleared on restart of docker.

In this solution I found that if the docker bridge rules are missing these will be inserted before the rules from the default iptables file. 

If you like I can make a complete example for the documentation.

**Comment by thaJeztah** [comment]
that would be appreciated! /cc @aboch

**Comment by jcstover** [comment]
I've created a default config file for iptables in which there is a FORWARD rule to a DOCKER-BLOCK chain to restrict access to docker containers. In the example there are the following assumptions:
* network 192.168.0.0/16 is given access to the host
* network 192.168.123.0/24 is given access to the containers
* docker networks 10.13.0.0/16 (our overlay network) and 172.18.0.0/16 are given acces to the containers
* normal bridge docker0 is present
* overlay bridge docker_gwbridge is present

Loading this iptables config in iptables and then start the docker daemon. Check if the FORWARD rules sending traffic to the DOCKER-BLOCK chain are present at top of the FORWARD chain. Mind that DOCKER-ISOLATION can be at the top.

File is at [https://gist.github.com/jcstover/135c082fb1a2d2f4b519d78c327852f4](iptables)

Note: it might be that the check of existing rules goes wrong in docker and that might be the reason why this solution works. The rules are in the correct order but the order is changed because I use insert for the BLOCKER-CHAIN Forward rule, so it ends on top. It may that the docker code is not checking if the rule is appended or inserted. But for me this is a workable solution.

**Comment by GameScripting** [comment]
I just (freshly on a new server) installed Docker 1.12 and noticed that the FORWARD chain is now set to DROP.

Is this indented? Anyone else to confirm that behaviour?

**Comment by jcstover** [comment]
I believe this was a wish as expressed in #14041. If you haven't enabled forwarding and docker enables it it used to forward all traffic and thus opening your containers to the outside world. By setting the FORWARD to drop, security whise nothing changes, and if you want to allow traffic you should add the needed rules.

**Comment by stszap** [comment]
@jcstover Thank you for your workaround, but I'm afraid that it isn't working for me.
What I tried:
start docker
list iptables rules
```
# iptables -L FORWARD -nv  --line-numbers
Chain FORWARD (policy ACCEPT 0 packets, 0 bytes)
num   pkts bytes target     prot opt in     out     source               destination         
1        0     0 DOCKER-ISOLATION  all  --  *      *       0.0.0.0/0            0.0.0.0/0           
2        0     0 DOCKER     all  --  *      docker0  0.0.0.0/0            0.0.0.0/0           
3        0     0 ACCEPT     all  --  *      docker0  0.0.0.0/0            0.0.0.0/0            ctstate RELATED,ESTABLISHED
4        0     0 ACCEPT     all  --  docker0 !docker0  0.0.0.0/0            0.0.0.0/0           
5        0     0 ACCEPT     all  --  docker0 docker0  0.0.0.0/0            0.0.0.0/0
```
stop docker
add test rule after DOCKER-ISOLATION
`iptables -I FORWARD 2 -s 8.8.8.8 -j ACCEPT`
list rules again
```
# iptables -L FORWARD -nv  --line-numbers
Chain FORWARD (policy ACCEPT 0 packets, 0 bytes)
num   pkts bytes target     prot opt in     out     source               destination         
1        0     0 DOCKER-ISOLATION  all  --  *      *       0.0.0.0/0            0.0.0.0/0           
2        0     0 ACCEPT     all  --  *      *       8.8.8.8              0.0.0.0/0           
3        0     0 DOCKER     all  --  *      docker0  0.0.0.0/0            0.0.0.0/0           
4        0     0 ACCEPT     all  --  *      docker0  0.0.0.0/0            0.0.0.0/0            ctstate RELATED,ESTABLISHED
5        0     0 ACCEPT     all  --  docker0 !docker0  0.0.0.0/0            0.0.0.0/0           
6        0     0 ACCEPT     all  --  docker0 docker0  0.0.0.0/0            0.0.0.0/0
```
start docker
list rules again
```
# iptables -L FORWARD -nv
Chain FORWARD (policy ACCEPT 0 packets, 0 bytes)
pkts bytes target     prot opt in     out     source               destination         
0     0 DOCKER-ISOLATION  all  --  *      *       0.0.0.0/0            0.0.0.0/0           
0     0 DOCKER     all  --  *      docker0  0.0.0.0/0            0.0.0.0/0           
0     0 ACCEPT     all  --  *      docker0  0.0.0.0/0            0.0.0.0/0            ctstate RELATED,ESTABLISHED
0     0 ACCEPT     all  --  docker0 !docker0  0.0.0.0/0            0.0.0.0/0           
0     0 ACCEPT     all  --  docker0 docker0  0.0.0.0/0            0.0.0.0/0           
0     0 ACCEPT     all  --  *      *       8.8.8.8              0.0.0.0/0
```

And my rule is at the bottom. I tried to insert in at the top, but it always ends at the bottom of FORWARD chain after docker is started. 

```
docker version
Client:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:        Thu Aug 18 05:13:43 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:        Thu Aug 18 05:13:43 2016
 OS/Arch:      linux/amd64
```

**Comment by GongT** [comment]
my custom rule always below of docker's rules. (jump to firewall-filter chain)

```
[gw.b.bj ~]# iptables -L
Chain INPUT (policy ACCEPT)
target     prot opt source               destination
firewall-filter  all  --  anywhere             anywhere

Chain FORWARD (policy ACCEPT)
target     prot opt source               destination
DOCKER-ISOLATION  all  --  anywhere             anywhere
DOCKER     all  --  anywhere             anywhere
ACCEPT     all  --  anywhere             anywhere             ctstate RELATED,ESTABLISHED
ACCEPT     all  --  anywhere             anywhere
ACCEPT     all  --  anywhere             anywhere
firewall-filter  all  --  anywhere             anywhere
```

how to prevent docker from modify FORWARD chain !!!!
```
[gw.b.bj ~]# docker info
Containers: 7
 Running: 7
 Paused: 0
 Stopped: 0
Images: 115
Server Version: 1.12.3
Storage Driver: devicemapper
 Pool Name: docker-253:1-402519-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: xfs
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 616.1 MB
 Data Space Total: 107.4 GB
 Data Space Available: 35.6 GB
 Metadata Space Used: 2.011 MB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.145 GB
 Thin Pool Minimum Free Space: 10.74 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 WARNING: Usage of loopback devices is strongly discouraged for production use. Use `--storage-opt dm.thinpooldev` to specify a custom block storage device.
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.107-RHEL7 (2016-06-09)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: overlay host bridge null
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: seccomp
Kernel Version: 3.10.0-327.36.3.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 993 MiB
Name: iz2zebgky0qb6npu5tkvz5z
ID: H3FY:PJEW:RGRU:E732:HLI3:FFZB:DUPA:LJO3:Q3B5:PK7P:JXJQ:4R63
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Insecure Registries:
 127.0.0.0/8
```

**Comment by clems4ever** [comment]
Hello,

+1 for a solution.
I'm also searching for a solution to route the packets to a custom chain (for filtering) via FORWARD before they go to the docker ones (DOCKER-INGRESS and DOCKER-ISOLATION). When I restart the docker engine, my rule that routes packets to the custom chain is unexpectedly going down in the list and my policy is not met anymore...

Same use case as @stszap.

**Comment by gw0** [comment]
+1

It is not just that Docker is updating the FORWARD chain (especially reinserting the DOCKER-ISOLATION rule on top) on startup, but also each time a container is started! If it would insert DOCKER-ISOLATION only once on startup, then a `ExecStartPost=` rules for Systemd could provide a way to work around Dockers limitations, but now such additions just get reset.

**Comment by clems4ever** [comment]
Here is my solution using cron: https://gist.github.com/clems4ever/55888f04ef78d519554add110aab126a

**Comment by gw0** [comment]
@clems4ever Sorry, but I do not see *cron* in your "solution". You are just using a systemd service that runs a bash script in an infinite loop in 10 seconds interval.

**Comment by clems4ever** [comment]
Hello @gw0, you are totally right, I changed my comment accordingly. Eventually I did not use cron because I wanted my rules to be applied as quickly as possible whereas cron has a 1-minute window at best.

**Comment by thaJeztah** [comment]
Reopening for now, because https://github.com/docker/libnetwork/issues/1675 was not yet vendored in this repository, so the code is not in use yet

ping @sanimej will there be another libnetwork bump before 17.06?

**Comment by thaJeztah** [comment]
Docker 17.06  (through commit https://github.com/moby/moby/pull/32981/commits/46392f2442137ad52bb7382e91e7ad4b70d323b9, which is part of https://github.com/moby/moby/pull/32981) includes a change to libnetwork (thanks @wenjianhn) that allows setting up a custom `"DOCKER-USER"` filter chain (see the change in libnetwork upstream: https://github.com/docker/libnetwork/pull/1675), so for those interested; upgrade to 17.06 to get this enhancement

**Comment by trapier** [comment]
@thaJeztah: this is great news.  Can a line regarding the `DOCKER-USER` chain be added to the 17.06 docker-ce release notes?  ( https://github.com/docker/docker-ce/releases/tag/v17.06.0-ce-rc1 )

**Comment by thaJeztah** [comment]
@trapier can you open a pull request in that repo (against the 17.06 branch; https://github.com/docker/docker-ce/blob/17.06/CHANGELOG.md)? We can update the changelog on GitHub with the same information afterwards
